filesource/filesource.go

```diff
@@ -125,100 +128,115 @@ type connector struct {
-	var result = &airbyte.ConnectionStatus{
-		Status: airbyte.StatusSucceeded,
+// Discover returns the set of resources available from this Driver.
+func (src *Source) Discover(ctx context.Context, req *pc.Request_Discover) (*pc.Response_Discovered, error) {
+	var cfg, err = src.NewConfig(req.ConfigJson)
+	if err != nil {
+		return nil, fmt.Errorf("parsing config json: %w", err)
 	}
 
-	var _, err = newConnector(src, args.ConfigFile)
+  store, err := src.Connect(ctx, cfg)
 	if err != nil {
-		result.Status = airbyte.StatusFailed
-		result.Message = err.Error()
+		return nil, fmt.Errorf("connecting to store: %w", err)
 	}
 
-	return airbyte.NewStdoutEncoder().Encode(airbyte.Message{
-		Type:             airbyte.MessageTypeConnectionStatus,
-		ConnectionStatus: result,
-	})
+	var conn = connector{config: cfg, store: store}
+
+  var root = conn.config.DiscoverRoot()
+  resourceJSON, err := json.Marshal(resource{Stream: root})
+
+	return &pc.Response_Discovered{Bindings: []*pc.Response_Discovered_Binding{{
+			RecommendedName:    pf.Collection(root),
+			ResourceConfigJson: resourceJSON,
+			DocumentSchemaJson: json.RawMessage(minimalDocumentSchema),
+			Key:                []string{"/_meta/file", "/_meta/offset"},
+  }}}, nil
 }
 
-func (src Source) Discover(args airbyte.DiscoverCmd) error {
-	var conn, err = newConnector(src, args.ConfigFile)
+// Pull is a very long lived RPC through which the Flow runtime and a
+// Driver cooperatively execute an unbounded number of transactions.
+func (src *Source) Pull(open *pc.Request_Open, stream *boilerplate.PullOutput) error {
+	var cfg, err = src.NewConfig(open.Capture.ConfigJson)
 	if err != nil {
-		return err
+		return fmt.Errorf("parsing config json: %w", err)
 	}
 
-	return airbyte.NewStdoutEncoder().Encode(airbyte.Message{
-		Type: airbyte.MessageTypeCatalog,
-		Catalog: &airbyte.Catalog{
-			Streams: []airbyte.Stream{{
-				Name:               conn.config.DiscoverRoot(),
-				JSONSchema:         json.RawMessage(minimalDocumentSchema),
-				SupportedSyncModes: airbyte.AllSyncModes,
-				SourceDefinedPrimaryKey: [][]string{
-					{"_meta", "file"},
-					{"_meta", "offset"},
-				},
-			}},
-		},
-	})
-}
+	var ctx = stream.Context()
 
-func (src Source) Read(args airbyte.ReadCmd) error {
-	var conn, err = newConnector(src, args.ConfigFile)
+  store, err := src.Connect(ctx, cfg)
 	if err != nil {
-		return err
-	}
-	var catalog = airbyte.ConfiguredCatalog{}
-	if err = args.CatalogFile.Parse(&catalog); err != nil {
-		return fmt.Errorf("parsing configured catalog: %w", err)
-	}
-	if catalog.Range.IsZero() {
-		// Process all files, unless the parsed catalog says otherwise.
-		catalog.Range = airbyte.NewFullRange()
+		return fmt.Errorf("connecting to store: %w", err)
 	}
 
-	var states = make(States)
-	if args.StateFile != "" {
-		if err = args.StateFile.Parse(&states); err != nil {
-			return fmt.Errorf("parsing state: %w", err)
+	var conn = connector{config: cfg, store: store}
+
+	var states States
```

Comment:
Does this work when `open.StateJson != nil` to unmarshall into? I've hit this before where there's a problem with the `nil` val vs. using `make` to initialize as an empty map.